lib/redisHandler.js

```diff
@@ -0,0 +1,231 @@
+
+/**
+ * Class to handle a redis connection and provide events to react on for
+ * all incoming Redis commands
+ */
+class RedisHandler extends EventEmitter {
+    /**
+     * Initialize and register all data handlers to send out events on new commands
+     * @param socket Network Socket/Connection
+     * @param options options objects, currently mainly for logger
+     */
+    constructor(socket, options) {
+        super();
+
+        options = options || {};
+        this.log = options.log || console;
+        this.socket = socket;
+
+        this.sockedId = socket.remoteAddress + ':' + socket.remotePort;
+        this.initialized = false;
+
+        this.writeQueue = [];
+
+        this.resp = new Resp();
+
+        this.resp.on('error', (err) => {
+            this.log.error(this.sockedId + ' (Init=' + this.initialized + ') Redis error:' + err);
+            if (this.initialized) {
+                this.sendError(null,new Error('PARSER ERROR ' + err)); // TODO
+            }
+            else {
+                this.close();
+            }
+        });
+
+        this.resp.on('data', (data) => {
+            const command = data.splice(0, 1)[0];
+            if (command === 'info') this.initialized = true;
+            const t = process.hrtime();
+            const responseId = (t[0] * 1e3) + (t[1] / 1e6);
+            this.log.silly(this.sockedId + ' Parser result: id=' + responseId + ', command=' + command + ', data=' + JSON.stringify(data)); // TODO remove data logging (performance)
+            this.writeQueue.push({id: responseId, data: false});
+            if (this.listenerCount(command) !== 0) {
+                this.emit(command, data, responseId);
+            }
+            else {
+                this.sendError(responseId, new Error(command + 'NOT SUPPORTED'));
+            }
+        });
+        this.resp.on('finish', () => {
+            this.log.debug(this.sockedId + ' Redis finish');
+        });
+
+        socket.on('data', (data) => {
+            if (!data) return;
+            this.log.silly(this.sockedId + ' New Redis request: ' + data.toString().replace(/[\r\n]*/g, '')); // TODO remove data logging (performance)
+            this.resp.write(data);
+        });
+
+        socket.on('error', (err) => {
+            this.log.error(this.sockedId + ' Redis error: ' + err);
+            this.socket.destroy();
+        });
+    }
+
+    /**
+     * Check if the response to a certain command can be send out directly or
+     * if it needs to wait till earlier responses are ready
+     * @param responseId ID of the response
+     * @param data Buffer to send out
+     * @private
+     */
+    _sendQueued(responseId, data) {
+        let idx = 0;
+        while (this.writeQueue.length && idx <= this.writeQueue.length && this.writeQueue[idx]) {
+            if (this.writeQueue[idx].id === responseId) {
+                this.writeQueue[idx].data = data;
+                if (idx > 0) break;
```

Comment:
Comment: why break? (It's clear to me now, but I don't know how it is in a year)